File: release-gates/src/test/java/com/tongsr/kaleido/release/ReleaseDossierCliTest.java
package com.tongsr.kaleido.release;

import static org.junit.Assert.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

public final class ReleaseDossierCliTest {
    @Rule public final TemporaryFolder temporary = new TemporaryFolder();

    @Test
    public void closesExactRecordsAndIndependentApprovals() throws Exception {
        var root = temporary.getRoot().toPath();
        var manifest = write(root, "manifest", "schema=KaleidoImmutableReleaseManifest.v1\n"
                + "asset.0.sha256=candidate\nverdict=PASS\n");
        var arguments = new ArrayList<String>();
        arguments.addAll(java.util.List.of("--output", root.resolve("dossier").toString(),
                "--manifest", manifest.toString(), "--manifest-signature-verified", "true"));
        for (var name : java.util.List.of("cache", "matrix.A3", "matrix.A4", "runtime.A3",
                "runtime.A4", "performance", "provenance", "documentation", "portal-dry-run")) {
            var record = write(root, name.replace('.', '-'),
                    "candidate.sha256=candidate\nverdict=PASS\n");
            arguments.addAll(java.util.List.of("--record", name + "|" + record));
        }
        var owner = write(root, "owner", "candidate.sha256=candidate\nreviewer.id=one\n"
                + "role=release-owner\ndecision=APPROVE\nsignatureVerified=true\n");
        var security = write(root, "security", "candidate.sha256=candidate\nreviewer.id=two\n"
                + "role=provenance-security-reviewer\ndecision=APPROVE\nsignatureVerified=true\n");
        arguments.addAll(java.util.List.of("--approval", owner.toString(),
                "--approval", security.toString()));
        ReleaseDossierCli.main(arguments.toArray(String[]::new));
        assertTrue(Files.readString(root.resolve("dossier")).endsWith("verdict=PASS\n"));
    }

    private static Path write(Path root, String name, String contents) throws Exception {
        return Files.writeString(root.resolve(name + ".properties"), contents);
    }
}

File: release-gates/src/test/kotlin/com/tongsr/kaleido/release/ReleaseDossierCliTest.kt
package com.tongsr.kaleido.release

import java.nio.file.Files
import java.nio.file.Path
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder

class ReleaseDossierCliTest {
    @Rule
    @JvmField
    val temporary = TemporaryFolder()

    @Test
    fun closesExactRecordsAndIndependentApprovals() {
        val root = temporary.root.toPath()
        val manifest = write(
            root,
            "manifest",
            "schema=KaleidoImmutableReleaseManifest.v1\nasset.0.sha256=candidate\nverdict=PASS\n",
        )
        val manifestSignature = write(root, "manifest-signature", "signature")
        val arguments = ArrayList<String>()
        arguments.addAll(
            listOf(
                "--output", root.resolve("dossier").toString(),
                "--manifest", manifest.toString(),
                "--manifest-signature", manifestSignature.toString(),
                "--manifest-signature-verified", "true",
            ),
        )
        for (name in listOf(
            "cache", "matrix.A3", "matrix.A4", "runtime.A3",
            "runtime.A4", "performance", "provenance", "documentation", "portal-dry-run",
        )) {
            val record = write(
                root,
                name.replace('.', '-'),
                "candidate.sha256=candidate\nverdict=PASS\n",
            )
            arguments.addAll(listOf("--record", "$name|$record"))
        }
        val owner = write(
            root,
            "owner",
            "candidate.sha256=candidate\nreviewer.id=one\n" +
                "role=release-owner\ndecision=APPROVE\nsignatureVerified=true\n" +
                "signer.fingerprint=${"A".repeat(40)}\n",
        )
        val security = write(
            root,
            "security",
            "candidate.sha256=candidate\nreviewer.id=two\n" +
                "role=provenance-security-reviewer\ndecision=APPROVE\nsignatureVerified=true\n" +
                "signer.fingerprint=${"B".repeat(40)}\n",
        )
        val ownerSignature = write(root, "owner-signature", "signature-one")
        val securitySignature = write(root, "security-signature", "signature-two")
        arguments.addAll(
            listOf(
                "--approval", owner.toString(),
                "--approval-signature", ownerSignature.toString(),
                "--approval", security.toString(),
                "--approval-signature", securitySignature.toString(),
            ),
        )
        ReleaseDossierCli.main(arguments.toTypedArray())
        val dossier = Files.readString(root.resolve("dossier"))
        assertTrue(dossier.startsWith("schema=KaleidoPrePublicationDossier.v1\n"))
        assertTrue(dossier.endsWith("verdict=PASS\n"))
    }

    @Test
    fun rejectsRecordsThatAreNotBoundToTheCandidate() {
        val root = temporary.newFolder("unbound-record").toPath()
        val manifest = write(
            root,
            "manifest",
            "schema=KaleidoImmutableReleaseManifest.v1\nasset.0.sha256=candidate\nverdict=PASS\n",
        )
        val manifestSignature = write(root, "manifest-signature", "signature")
        val arguments = ArrayList<String>()
        arguments.addAll(
            listOf(
                "--output", root.resolve("dossier").toString(),
                "--manifest", manifest.toString(),
                "--manifest-signature", manifestSignature.toString(),
                "--manifest-signature-verified", "true",
            ),
        )
        for (name in listOf(
            "cache", "matrix.A3", "matrix.A4", "runtime.A3",
            "runtime.A4", "performance", "provenance", "documentation", "portal-dry-run",
        )) {
            val contents = if (name == "documentation") {
                "verdict=PASS\n"
            } else {
                "candidate.sha256=candidate\nverdict=PASS\n"
            }
            val record = write(root, name.replace('.', '-'), contents)
            arguments.addAll(listOf("--record", "$name|$record"))
        }
        val owner = write(
            root,
            "owner",
            "candidate.sha256=candidate\nreviewer.id=one\n" +
                "role=release-owner\ndecision=APPROVE\nsignatureVerified=true\n" +
                "signer.fingerprint=${"A".repeat(40)}\n",
        )
        val security = write(
            root,
            "security",
            "candidate.sha256=candidate\nreviewer.id=two\n" +
                "role=provenance-security-reviewer\ndecision=APPROVE\nsignatureVerified=true\n" +
                "signer.fingerprint=${"B".repeat(40)}\n",
        )
        val ownerSignature = write(root, "owner-signature", "signature-one")
        val securitySignature = write(root, "security-signature", "signature-two")
        arguments.addAll(
            listOf(
                "--approval", owner.toString(),
                "--approval-signature", ownerSignature.toString(),
                "--approval", security.toString(),
                "--approval-signature", securitySignature.toString(),
            ),
        )

        val failure = assertThrows(IllegalArgumentException::class.java) {
            ReleaseDossierCli.main(arguments.toTypedArray())
        }
        assertTrue(failure.message!!.contains("missing property: candidate.sha256"))
    }

    companion object {
        private fun write(root: Path, name: String, contents: String): Path =
            Files.writeString(root.resolve("$name.properties"), contents)
    }
}
